File: src/ragapp/admin/client/tools/image_generator.ts
import { z } from "zod";

export const ImageGeneratorToolConfig = z.object({
  name: z.literal("image_generator"),
  label: z.string().nullable().optional(),
  description: z.string().nullable().optional(),
  enabled: z.boolean().nullable().optional(),
  config: z
    .object({
      api_key: z
        .string()
        .nullable()
        .refine(
          (data) => {
            return data && data.length > 0;
          },
          { message: "API Key is required" },
        ),
    })
    .nullable()
    .optional(),
});
export type ImageGeneratorToolConfigType = z.infer<
  typeof ImageGeneratorToolConfig
>;
export const DEFAULT_IMAGE_GENERATOR_TOOL_CONFIG = {
  label: "Image Generator",
  description:
    "Generate images from the provided text using the Stability AI API",
  config: {
    api_key: "",
  },
  enabled: false,
};

File: src/ragapp/backend/models/tools/image_generator.py
from typing import ClassVar, Literal, Optional

from pydantic import BaseModel, Field


class ImageGeneratorToolConfig(BaseModel):
    api_key: str | None = Field(
        default=None,
        description="The Stability AI API key to use for the image generator.",
    )


class ImageGeneratorTool(BaseModel):
    config_id: ClassVar[str] = "img_gen"
    name: Literal["ImageGenerator"] = "ImageGenerator"
    tool_type: Literal["local"] = "local"
    label: Literal["Image Generator"] = "Image Generator"
    description: str = "Generate images from text descriptions."
    config: ImageGeneratorToolConfig | None = Field(
        default=ImageGeneratorToolConfig(),
    )
    custom_prompt: Optional[str] = (
        """The output link of the generated image must start with '/api/files/output/tool/<image>'. It must be a relative path, don't try adding the domain name or protocol"""
    )
    enabled: bool = False

    def validate_config(self) -> bool:
        if self.enabled and not self.config.api_key:
            raise ValueError("API key is required for enabled Image Generator tool")
        return True
